### PR TITLE
Add meaningful string representations for resolution invariants

### PR DESCRIPTION
**Pull request by ecordell**
**Description of the change:**

Original description:
> During dependency resolution, invariants are enforced for
one-operator-per-package and one-provider-per-api. Their user-facing
string representation (shown when they prevent resolution from
succeeding) had re-used the same representation as constraints on
Subscriptions out of convenience.

> Recently, the messaging of Subscription-related constraints was
changed to be more descriptive which inadvertently made the invariant
messaging misleading. This patch improves constraint messaging for all
dependency relationships except bundle dependencies.

This rebase of #1994 adds additional tests when multiple subscriptions 
exist for the same package. Those tests uncovered a weird UX for an
edge case: if two subscriptions are present for the same package, and 
the resolver manages to find an operator that satisfies both.

This is not likely or common today, but since it is a potential source of
confusion I updated the resolver to ensure that in that case, both 
subscriptions are updated to point to the operator that satisfies both.

Prior to that change, only one of the subscriptions that matched would 
be picked (whichever came first in the query).

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage 
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/docs` 
- [ ] Commit messages sensible and descriptive

### CODE REVIEW

File: pkg/controller/registry/resolver/installabletypes.go [hunk 1]
```diff
@@ -62,58 +61,100 @@ func NewBundleInstallable(bundle, channel string, catalog registry.CatalogKey, c
...
+	} else {
+		req = fmt.Sprintf("at least one of %s or %s", strings.Join(s[:len(s)-1], ", "), s[len(s)-1])
+	}
+	result.constraints = append(result.constraints, PrettyConstraint(solver.Dependency(dependencies...), fmt.Sprintf("subscription %s requires %s", name, req)))
+
+	return result
+}
+
+func NewSingleAPIProviderInstallable(group, version, kind string, providers []solver.Identifier) solver.Installable {
+	gvk := fmt.Sprintf("%s (%s/%s)", kind, group, version)
```
Comment (kevinrizza) [inline pkg/controller/registry/resolver/installabletypes.go:106]: nit: this seems like something that should be factored out if possible

File: pkg/controller/registry/resolver/installabletypes.go [hunk 2]
```diff
@@ -62,58 +61,100 @@ func NewBundleInstallable(bundle, channel string, catalog registry.CatalogKey, c
...
+	result := GenericInstallable{
+		identifier: solver.IdentifierFromString(pkg),
+	}
+	if len(providers) <= 1 {
+		// The constraints are pointless without more than one provider.
+		return result
+	}
+	result.constraints = append(result.constraints, PrettyConstraint(solver.Mandatory(), fmt.Sprintf("there can be only one operator from package %s", pkg)))
+
+	var s []string
```
Comment (kevinrizza) [inline pkg/controller/registry/resolver/installabletypes.go:136]: super  nit: `providersString` maybe? This loop would probably be nicer as a closure but there's not good syntactic sugar to do that in go

File: pkg/controller/registry/resolver/resolver.go [hunk 3]
```diff
@@ -449,12 +448,7 @@ func (r *SatResolver) addInvariants(namespacedCache MultiCatalogOperatorFinder,
 			if err != nil {
 				continue
 			}
-			key := fmt.Sprintf("gvkunique/%s/%s/%s", prop.Group, prop.Version, prop.Kind)
-			_, ok := conflictToInstallable[key]
-			if !ok {
-				conflictToInstallable[key] = make([]solver.Installable, 0)
-			}
-			conflictToInstallable[key] = append(conflictToInstallable[key], installable)
+			gvkConflictToInstallable[prop] = append(gvkConflictToInstallable[prop], installable.Identifier())
```
Comment (kevinrizza) [inline pkg/controller/registry/resolver/resolver.go:451]: nice

**Review by kevinrizza — COMMENTED** [review]
/lgtm

**Comment by ecordell** [comment]
/hold

The failure in the unit tests looks unrelated to this change, but it seems to happen consistently

**Comment by dinhxuanvu** [comment]
The unit test passed locally for me
